### DISCUSSION TITLE
Is handling of bNodes between Comunica and TPF Servers underspecified?

### DISCUSSION THREAD

**Discussion by larsgsvensson** [comunica/comunica#375] [Q&A] [answered]
#### Issue type:

- :question: Question

____
#### Question:

The [TPF specification §3](http://www.hydra-cg.com/spec/latest/triple-pattern-fragments/#data) says that "Triples SHOULD NOT contain blank nodes (if the original triples contain blank nodes, they SHOULD be skolemized)." This means that it is legal for a TPF Server to return data that contains blank nodes, e. g.
```turtle
[] a skos:Concept .
```
There have been several issues (latest one #361) saying that SPARQL queries don't work due to non-matching of blank nodes. Could this be because if a TPF server does not skolemize blank nodes, there simply is no way for the client and the server to ensure that they refer to the same blank node?
As an example, consider the following data:
```turtle
my:dataset a dcat:Dataset ;
    dcat:distribution [
        a dcat:Distribution ;
        dcat:downloadURL <https://example.org/data.mrcx> 
] .
my:otherDataset a dcat:Dataset ;
    dcat:distribution [
        a dcat:Distribution ;
        dcat:downloadURL <https://example.org/moreData.mrcx> 
] .
```
Now I want to find the `dcat:downloadURL`s for all distributions of a dataset:
```sparql
SELECT ?datasetURI ?downloadURL
WHERE {
    ?datasetURI a dcat:Dataset ;
        dcat:distribution ?dist .
    ?dist dcat:downloadURL ?downloadURL .
}
```
In this case the TPF server will try to find fragments such as
```turtle
subject=? predicate=dcat:downloadURL object=?
```
and will get back
```turtle
[] dcat:downloadURL <https://example.org/data.mrcx> .
[] dcat:downloadURL <https://example.org/moreData.mrcx> .
```
However, in my understanding there is no way that the client will know which distribution the download URL belongs to and thus it cannot join it to the correct dataset. Is this correct?

So this boils down to the question, if Comunica requires TPF servers to use skolemized blank nodes in the data they return? If so, that would mean that a server that uses the liberty not to use skolemized blank nodes (the spec says SHOULD) cannot be used as a data source for Comunica.

I'd love to be proven wrong!

**Comment by rubensworks** [comment]
Interesting question!

I think the main reason why the spec say that bnodes SHOULD NOT be used, instead of MUST NOT, is because there should not be any problems when using bnodes inside the _metadata_ part of the responses. But perhaps @RubenVerborgh can confirm this.

That being said, you're right about the danger of having blank nodes inside the actual _data_ part of the response, as blank nodes only make sense in the scope of single documents. This is something we intend to resolve in the future (#355).

In any case, all TPF server implementations I know of properly skolemize blank nodes. So #361 is probably not being caused by this, I assume the problem lies in Comunica itself (to be checked).

**Comment by RubenVerborgh** [comment, marked as answer]
I have updated the spec to say that data triples MUST NOT contain blank nodes.

**Comment by rubensworks** [comment]
Great! I guess that resolves this issue :-)

**Comment by larsgsvensson** [comment]
> Great! I guess that resolves this issue :-)

Yes, and it also makes all implementations of Server.java non-conformant, at least until https://github.com/LinkedDataFragments/Server.Java/issues/6 is solved.

**Comment by larsgsvensson** [comment]
@RubenVerborgh 

> I have updated the spec to say that data triples MUST NOT contain blank nodes.

Does that mean that if the original data contains blank nodes, they MUST be skolemized?

**Comment by RubenVerborgh** [comment]
They SHOULD be skolemized (https://github.com/HydraCG/Specifications/commit/7b4d4ab9ae21ba5d17b0ce0d5137663eb48fbed6); there are/might be other ways of making blanks disappear.

**Comment by larsgsvensson** [comment]
bq. ... there are/might be other ways of making blanks disappear.

Then I'd propose to phrase that more explicitly. Suggestion: "The RECOMMENDED way of removing blank nodes in data triples is <a href="http://www.w3.org/TR/rdf11-concepts/#section-skolemization">skolemization</a>."

**Comment by RubenVerborgh** [comment]
Thanks @larsgsvensson, done in https://github.com/HydraCG/Specifications/commit/d248173ad1539b59ab801bedfb3a51567016eee7.

**Comment by larsgsvensson** [comment]
OK, thanks @RubenVerborgh . Now we just need to make Server.java compliant again but that's a job for https://github.com/LinkedDataFragments/Server.Java/issues/6.
